# LIBRERÍA DE NÚMEROS COMPLEJOS 

En el presente documento, vamos a describir funciones que nos permitirán trabajar con números complejos, dentro del código de cada una de estas funciones realizadas en Python, se presenta una corta descripción de su funcionamiento como comentario; seguido a esto se encontrará una prueba para cada una de las funciones demostrando su funcionamiento y la solución del problema ingresado paso a paso para corroborar el resultado dado por la función.

## Representación de un número complejo
A lo largo de este documento, los números complejos de la forma $a+ib$ serán representados mediante listas de la siguiente manera: $[a,b]$.

Del mismo modo tenga en cuenta que 'c_1' y 'c_2' representan números complejos y que 'p_1' representa un número complejo en su forma polar.

### Suma

In [145]:
def suma(c_1, c_2):
    """
    La funsión suma recibe dos números complejos: c_1 y c_2 (Que deben ser listas de longitud 2) y retorna un complejo
    (lista con longitud 2) correspondiente a la operación c_1 + c_2.
    """
    return [c_1[0] + c_2[0], c_1[1] + c_2[1]]

Probemos la funsión ```suma```:

In [146]:
c_1 = [3,2]
c_2 = [7,5]
suma (c_1, c_2)

[10, 7]

Justificación de la prueba:

$ (3 + 2i) + (7 + 5i) $

$ = ((3 + 7) + (2 + 5)i) $

$ = (10 + 7i) $

### Resta

In [147]:
def resta(c_1, c_2):
    """
    La funsión suma recibe dos números complejos, c_1 y c_2 (Que deben ser listas de longitud 2) y retorna un complejo
    (lista con longitud 2) correspondiente a la operación c_1 - c_2.
    """
    return [c_1[0] - c_2[0], c_1[1] - c_2[1]]

Probemos la función ```resta```:

In [148]:
c_1 = [4,8]
c_2 = [3,1]
resta (c_1, c_2)

[1, 7]

Justificación de la prueba:

$ (4 + 8i) - (3 + 1i) $

= $ ((4 - 3) + (8 - 1)i) $

= $ (1 + 7i) $

### Producto

In [149]:
def producto(c_1, c_2):
    """
    La funsión producto recibe 2 números complejos, c_1 y c_2 (Que deben ser listas de longitud 2) y retorna un complejo
    (lista con longitud 2) correspondiente a la operación c_1 * c_2.
    """
    return [c_1[0] * c_2[0] - c_1[1] * c_2[1] , c_1[0] * c_2[1] + c_1[1] * c_2[0]]

Probemos la función ```producto```:

In [150]:
c_1 = [4,5]
c_2 = [-8,3]
producto (c_1, c_2)

[-47, -28]

Justificación de la prueba:
    
$ (4 + 5i) * (-8 + 3i) $

$ = -32 + 12i -40i + 15i^2 $

$ = -32 - 28i - 15 $

$ = (-47 -28i) $

### División

In [151]:
def division(c_1, c_2):
    """
    La funsión division recibe 2 números complejos, c_1 y c_2 (Que deben ser listas de longitud 2) y retorna un complejo
    (lista con longitud 2) correspondiente a la operación c_1 / c_2.
    """
    a = producto(c_1, conjugado(c_2))
    b = producto(c_2, conjugado(c_2))
    print(a, "/", b)
    return [a[0] / b[0], a[1] / b[0]]
    

Probemos la función ```division```:

In [152]:
c_1 = [4,-3]
c_2 = [2,5]
division(c_1, c_2)

[-7, -26] / [29, 0]


[-0.2413793103448276, -0.896551724137931]

Justificación de la prueba:

$ \frac{4 - 3i}{2 + 5i} $

$ = \frac{(4 - 3i) * (2 - 5i)}{(2 + 5i) * (2 - 5i)} $

$ = \frac{8 - 20i - 6i + 15i^2}{(4 - 10i + 10i - 25i^2} $

$ = \frac{(-7 - 26i)}{29} $

$ = \frac{-7}{29} \frac{-26}{29i} $

$ = (-0.24138 -0.89655i) $

### Conjugado

In [153]:
def conjugado(c_1):
    """
    La función conjugado recibe un número complejo, c_1 (que debe ser una lista de longitud 2) y retorna este mismo número, 
    pero con signo opuesto en la parte imaginaria.

    """
    return [c_1[0], c_1[1] * -1]

Probemos la función ```conjugado```:

In [154]:
c_1 = [-2,8]
conjugado(c_1)

[-2, -8]

Justificación de la prueba:

Se llama conjugado de un número complejo al número complejo que se obtiene por simetría del dado respecto del eje de abscisas, para el caso del número $ (-2,8i) $ se tiene como cojugado a $(-2,-8i)$.

### Módulo

In [155]:
import math
def modulo(c_1):
    """
    La funsión modulo recibe un número complejo, c_1 (que debe ser una lista de longitud 2) y retorna un número de tipo real.
    """
    rta = math.sqrt(c_1[0] ** 2 + c_1[1] ** 2)
    return rta

Probemos la función ```modulo```:

In [156]:
c_1 = [3,4]
modulo(c_1)

5.0


Justificación de la prueba:  

Módulo de $(3 + 4i) = \sqrt{3^2 + 4^2} = \sqrt{9 + 16} = \sqrt{25} = 5$

### Imprimir en la forma $ a + bi $

In [157]:
def imprimir1(c_1):
    """
    La funsión imprimir1 recibe un número complejo representado de la forma [a,b] y lo devuelve en la forma a + bi.
    """
    print(str(c_1[0]) + " + " + str(c_1[1]) + "i")

Probemos la función ```imprimir1```:

In [158]:
c_1 = [-1,5]
imprimir1(c_1)

-1 + 5i


### Convertir de forma cartesiana a polar

In [159]:
def cartesiana_polar(c_1):
    """
    La funsión cartesiana_polar recibe un número complejo c_1 en su forma binómica y lo retorna en su forma polar.
    En el comentario siguiente (precedido por el simbolo #), encontrará otra opcion de return que devolverá el ángulo en grados.
    """
    r = modulo(c_1)
    a = math.atan(c_1[1]/c_1[0])
    #return [r,math.degrees(a)] #(Ángulo en grados)
    return [r, a]

Probemos la función ```cartesiana_polar```:

In [160]:
c_1 = [2,2]
cartesiana_polar(c_1)

[2.8284271247461903, 0.7853981633974483]

Justificación de la prueba:

Radio $= \sqrt{2^2 + 2^2} = \sqrt{4 + 4} = \sqrt{8} = 2.828427$

Angulo $= atan(\frac{2}{2}) = atan(1) = 0.785398$

$ (2.828427 + 0.785398i) $

### Convertir de forma polar a cartesiana

In [161]:
def polar_cartesiana(p_1):
    """
    La funsión polar_cartesiana recibe un número complejo p_1 (que debe ser una lista de longitud 2) en su forma polar y 
    lo retorna en su forma binómica (lista con longitud 2).
    
    """
    a = p_1[0] * math.cos(p_1[1])
    b = p_1[0] * math.sin(p_1[1])
    return [a,b]

Probemos la función ```polar_cartesiana```:

In [162]:
p_1 = [2, math.pi/4]
polar_cartesiana(p_1)

[1.4142135623730951, 1.4142135623730951]

 Justificación de la prueba:
 
 Para pasar un número complejo de su forma polar a su forma cartesiana utilizaremos la siguiente fórmula: 

$$ (a + bi) = (r * cos(α) + r * sin(α)i) $$
 
 donde $r$ es el radio y $α$ el ángulo, así:

$ (a + bi) = (2 * cos(\frac{pi}{4}) + 2 * sin(\frac{pi}{4})i) $

$ (a + bi) = (2 * \frac{\sqrt{2}}{2} + 2 * \frac{\sqrt{2}}{2}i) $

$ (a + bi) = (\sqrt{2} + \sqrt{2}i) $

$ (a + bi) = (1.41421, 1.41421i) $

### Imprimir en forma polar - exponencial

In [163]:
def imprimir_exp(p_1):
    """
    La funsión imprimir_exp recibe un número complejo p_1 (que debe ser una lista de longitud 2) en su forma polar y devuelve 
    este número en su forma exponencial.
    """
    print(str(p_1[0]) + "e^(i" + str(round(p_1[1],4)) + ")")

Probemos la función ```imprimir_exp```:

In [164]:
p_1 = [2, math.pi]
imprimir_exp(p_1)

2e^(i3.1416)


### Potencia n de un número complejo

In [165]:
def potencia_n(c_1, n):
    """
    la funsión recibe un número complejo c_1 en su forma cartesiana y devuelve este mismo elevado a una potencia n.
    """
  #Ingresando un número complejo en su forma binómica
    c_p_1 = cartesiana_polar (c_1)
    print("Forma Polar:")
    print(c_p_1)
    rta = polar_cartesiana([c_p_1[0] ** n, c_p_1[1] * n])
    return [round(rta[0]), round(rta[1])]

Probemos la función ```potencia_n```:

In [166]:
c_1 = [1, 1]
n = 4
potencia_n(c_1, n)

Forma Polar:
[1.4142135623730951, 0.7853981633974483]


[-4, 0]

 Justificación de la prueba: 

Comenzaremos hallando la forma polar del número complejo $(1 + i)$ que se encuentra en su forma binómica:

Radio $= \sqrt{1^2 + 1^2} = \sqrt{1 + 1} = \sqrt{2} = 1.414213$

Angulo $= atan(\frac{1}{1}) = atan(1) = 0.785398$

A partir de la forma polar del número dado, utilizamos la siguiente fórmula: 

$$(r, α)^n = (r^n, α * n)$$

donde $r$ es el radio, $α$ es el ángulo y $n$ es la potencia a elevar, así:

$(\sqrt{2}, atan(1))^4 = (\sqrt{2}^4, atan(1) * 4) = (4, pi)$

Lluego delvovemos el número elevado a la potencia 4 a su forma binómica:


$ (a + bi) = (4 * cos(pi) + 4 * sin(pi)i) $

$ (a + bi) = (4 * (-1) + 2 * 0i) $

$ (a + bi) = (-4 + 0i) $